File: 백준/Silver/15649. N과 M （1）/N과 M （1）.cc
#include <iostream>
#include <algorithm>

using namespace std;

int N, M;
int result[9];
int visited[9] = { 0, };

void DFS(int cnt)
{
	if (cnt == M)
	{
		for (int i = 0; i < M; i++)
		{
			cout << result[i] << " ";
		}
		cout << '\n';
		return;
	}

	for (int i = 1; i <= N; i++)
	{
		if (!visited[i])
		{
			visited[i] = 1; // true
			result[cnt] = i;
			DFS(cnt + 1);
			visited[i] = 0; // false
		}
	}
}

int main()
{
	cin >> N >> M;
	DFS(0);
	return 0;
}

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.java
import java.util.*;
import java.io.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;

    static int n, m;
    static int[] result;
    static boolean[] visited;

    public static void main(String[] args) throws Exception{
        st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        result = new int[m];
        visited = new boolean[n+1];

        dfs(0);
    }

    static void dfs(int cur) {

        if (cur == m) {
            for (int i = 0; i < result.length ; i ++) {
                System.out.print(result[i] + " ");
            }
            System.out.println();

            return;
        }

        for (int i = 1 ; i <= n ; i ++) {
            if (!visited[i]) {
                result[cur] = i;
                visited[i] = true;
                dfs(cur + 1);
                visited[i] = false;
            }
        }
    }

}